File: python/tree/235. Lowest Common Ancestor of a Binary Search Tree.py
'''
Time: O(logn) -- height of the tre
Space: O(1)
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # Iteration
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        cur = root  # current

        while cur:  # if cur != None
            if p.val < cur.val and q.val < cur.val:
                cur = cur.left    
            elif p.val > cur.val and q.val > cur.val:
                cur = cur.right
            else:
                return cur
            
    # Recursion   
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
            if p.val < root.val and q.val < root.val:
                return self.lowestCommonAncestor(root.left, p, q)   
            elif p.val > root.val and q.val > root.val:
                return self.lowestCommonAncestor(root.right, p, q)   
            else:
                return root

File: javascript/tree/235. Lowest Common Ancestor of a Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */

// Iteration
var lowestCommonAncestor = function(root, p, q) {
    let cur = root;

    while (cur) {
        if (p.val > cur.val && q.val > cur.val) cur = cur.right;
        else if (p.val < cur.val && q.val < cur.val) cur = cur.left;
        else return cur;
    }
};
// Recursion
var lowestCommonAncestor = function(root, p, q) {
    if (p.val > root.val && q.val > root.val) return lowestCommonAncestor(root.right, p, q);
    else if (p.val < root.val && q.val < root.val) return lowestCommonAncestor(root.left, p, q);
    else return root;
};